### PR TITLE
OJ-2545: Add CloudFront header tests

### PR DESCRIPTION
**Pull request by MarieseMikely**
**OJ-2545: Add CloudFront integration tests**

**Tests**
- Check for the device information header in the CRI start message
- Use the SQS and CloudFormation helpers from common-lib
- Update GHA to run the new integration tests
   Run integration tests sequentially between PRs to avoid SQS concurrency issues.
- Use `JsonPointer` to make assertions for JSON objects
- Upgrade the version of the common-lib to get the SQS and CF helper
  Update the AWS SDK version to match.

**Fixes**
- Use prefixed TxMA exports
  The non-prefixed exports are deprecated and will be removed in the future.
- Use the correct secret prefix for local developer stacks

### CODE REVIEW

File: integration-tests/src/test/java/gov/uk/address/api/stepdefinitions/AddressSteps.java [hunk 1]
```diff
@@ -1,5 +1,18 @@
 package gov.uk.address.api.stepdefinitions;
 
+import com.amazonaws.regions.Regions;
+import com.amazonaws.services.cloudformation.AmazonCloudFormation;
+import com.amazonaws.services.cloudformation.AmazonCloudFormationClientBuilder;
+import com.amazonaws.services.cloudformation.model.DescribeStacksRequest;
+import com.amazonaws.services.cloudformation.model.Stack;
+import com.amazonaws.services.cloudformation.model.*;
```
Comment (jkunle) [inline integration-tests/src/test/java/gov/uk/address/api/stepdefinitions/AddressSteps.java:8]: Please could you make this explicit, generally we avoid * imports

File: integration-tests/src/test/java/gov/uk/address/api/stepdefinitions/AddressSteps.java [hunk 2]
```diff
@@ -1,5 +1,18 @@
...
+
+import com.amazonaws.services.sqs.AmazonSQS;
+import com.amazonaws.services.sqs.AmazonSQSClientBuilder;
+import com.amazonaws.services.sqs.model.*;
```
Comment (jkunle) [inline integration-tests/src/test/java/gov/uk/address/api/stepdefinitions/AddressSteps.java:12]: same here, please get rid of * import

File: integration-tests/src/test/java/gov/uk/address/api/stepdefinitions/AddressSteps.java [hunk 3]
```diff
@@ -15,9 +28,7 @@
 import java.text.ParseException;
 import java.util.List;
 
-import static org.junit.jupiter.api.Assertions.assertEquals;
-import static org.junit.jupiter.api.Assertions.assertNotNull;
-import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.junit.jupiter.api.Assertions.*;
```
Comment (jkunle) [inline integration-tests/src/test/java/gov/uk/address/api/stepdefinitions/AddressSteps.java:31]: Here as well

Comment (jkunle) [reply integration-tests/src/test/java/gov/uk/address/api/stepdefinitions/AddressSteps.java:31]: see https://gds-way.digital.cabinet-office.gov.uk/manuals/programming-languages/java.html#imports

**Review by jkunle — APPROVED** [review]

**Review by jkunle — DISMISSED** [review]

File: .github/workflows/check-pr.yml [hunk 4]
```diff
@@ -27,11 +27,15 @@ jobs:
     name: Run tests
     needs: deploy
     uses: ./.github/workflows/run-integration-tests.yml
+    permissions:
```
Comment (Caitlin-cooling) [inline .github/workflows/check-pr.yml:30]: Just sense checking what these added permissions are for?

File: infrastructure/lambda/template.yaml [hunk 5]
```diff
@@ -386,7 +390,11 @@ Resources:
       Environment:
         Variables:
           POWERTOOLS_SERVICE_NAME: !Sub "${CriIdentifier}-postcode-lookup"
-          SQS_AUDIT_EVENT_QUEUE_URL: !ImportValue AuditEventQueueUrl
+          SQS_AUDIT_EVENT_QUEUE_URL:
+            Fn::ImportValue:
+              !Sub
+                - ${stackName}-AuditEventQueueUrl
+                - stackName: !If [IsDevEnvironment, !Ref CommonStackName, !Ref TxmaStackName]
```
Comment (Caitlin-cooling) [inline infrastructure/lambda/template.yaml:397]: Do when @CharlesIC has moved the queues into the txma-infra repo we should be able to remove this, right? We will always be using the `TxmaStackName`?

File: integration-tests/src/test/java/gov/uk/address/api/stepdefinitions/AddressSteps.java [hunk 6]
```diff
@@ -8,27 +8,41 @@
...
+    private ObjectMapper objectMapper;
     private final AddressApiClient addressApiClient;
     private final CriTestContext testContext;
     private String uprn;
     private String postcode;
 
+    private final String auditEventQueueName =
+            CloudFormationHelper.getOutput(
+                    CloudFormationHelper.getParameter(
+                            System.getenv("STACK_NAME"), "CommonStackName"),
```
Comment (Caitlin-cooling) [inline integration-tests/src/test/java/gov/uk/address/api/stepdefinitions/AddressSteps.java:38]: Right so this is going to have to change if we put the mock queues into the txma infra stack too?

File: integration-tests/src/test/java/gov/uk/address/api/stepdefinitions/AddressSteps.java [hunk 7]
```diff
@@ -112,4 +162,14 @@ public void user_does_not_get_any_address() throws JsonProcessingException {
         assertTrue(list.isEmpty());
         assertEquals(200, this.testContext.getResponse().statusCode());
     }
+
+    @And("the SQS events are deleted from the queue")
```
Comment (Caitlin-cooling) [inline integration-tests/src/test/java/gov/uk/address/api/stepdefinitions/AddressSteps.java:166]: This is nice to have and will let us know if we have any cleanup issues like we've been seeing in NINo 👍

File: .github/workflows/check-pr.yml (hunk 4 above)
Comment (MarieseMikely) [reply .github/workflows/check-pr.yml:30]: The permissions are needed to be able to get the required parameters and outputs from the stack without updating it in the template.yaml

File: infrastructure/lambda/template.yaml (hunk 5 above)
Comment (MarieseMikely) [reply infrastructure/lambda/template.yaml:397]: Yes, that's correct

File: integration-tests/src/test/java/gov/uk/address/api/stepdefinitions/AddressSteps.java (hunk 6 above)
Comment (MarieseMikely) [reply integration-tests/src/test/java/gov/uk/address/api/stepdefinitions/AddressSteps.java:38]: Yeah that'll change after @CharlesIC PR is merged

File: integration-tests/src/test/java/gov/uk/address/api/stepdefinitions/AddressSteps.java [hunk 8]
```diff
@@ -8,27 +8,41 @@
...
+import java.util.Collections;
 import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
 
-import static org.junit.jupiter.api.Assertions.assertEquals;
-import static org.junit.jupiter.api.Assertions.assertNotNull;
-import static org.junit.jupiter.api.Assertions.assertTrue;
+import static java.util.Map.entry;
+import static org.junit.jupiter.api.Assertions.*;
```
Comment (jkunle) [inline integration-tests/src/test/java/gov/uk/address/api/stepdefinitions/AddressSteps.java:26]: Please address wild card import here `import static org.junit.jupiter.api.Assertions.*;`

**Review by Caitlin-cooling — DISMISSED** [review]

**Review by CharlesIC — APPROVED** [review]

**Pull request merged**